packages/x-date-pickers/src/internals/hooks/useField/useField.ts

```diff
@@ -0,0 +1,415 @@
+    },
+    valueManager,
+    fieldValueManager,
+    validator,
+  } = params;
+
+  const firstDefaultValue = React.useRef(defaultValue);
+
+  const valueParsed = React.useMemo(() => {
+    // TODO: Avoid this type casting, the emptyValues are both valid TDate and TInputDate
+    const value = firstDefaultValue.current ?? (valueManager.emptyValue as unknown as TInputValue);
+
+    return valueManager.parseInput(utils, value);
+  }, [valueManager, utils]);
+
+  const [state, setState] = React.useState<UseFieldState<TValue, TSection[]>>(() => {
+    const sections = fieldValueManager.getSectionsFromValue(utils, null, valueParsed, format);
+
+    return {
+      sections,
+      valueParsed,
+      valueStr: fieldValueManager.getValueStrFromSections(sections),
+      selectedSectionIndexes: null,
+    };
+  });
+
+  const updateSections = (sections: TSection[]) => {
+    const { value: newValueParsed, shouldPublish } = fieldValueManager.getValueFromSections(
+      utils,
+      state.sections,
+      sections,
+      format,
+    );
+
+    setState((prevState) => ({
+      ...prevState,
+      sections,
+      valueStr: fieldValueManager.getValueStrFromSections(sections),
+      valueParsed: newValueParsed,
+    }));
+
+    if (onChange && shouldPublish) {
+      onChange(newValueParsed);
+    }
+  };
+
+  const updateSelectedSections = (start?: number, end?: number) => {
+    setState((prevState) => ({
+      ...prevState,
+      selectedSectionIndexes: start == null ? null : { start, end: end ?? start },
+      selectedSectionQuery: null,
+    }));
+  };
+
+  const handleInputClick = useEventCallback(() => {
+    if (state.sections.length === 0) {
+      return;
+    }
+
+    const nextSectionIndex = state.sections.findIndex(
+      (section) => section.start > (inputRef.current?.selectionStart ?? 0),
+    );
+    const sectionIndex = nextSectionIndex === -1 ? state.sections.length - 1 : nextSectionIndex - 1;
+
+    updateSelectedSections(sectionIndex);
+  });
+
+  const handleInputKeyDown = useEventCallback((event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (!inputRef.current || state.sections.length === 0) {
+      return;
+    }
+
+    // eslint-disable-next-line default-case
+    switch (true) {
+      // Select all
+      case event.key === 'a' && event.ctrlKey: {
+        event.preventDefault();
+        updateSelectedSections(0, state.sections.length - 1);
+        return;
+      }
+
+      // Move selection to next section
+      case event.key === 'ArrowRight': {
+        event.preventDefault();
+
+        if (state.selectedSectionIndexes == null) {
+          updateSelectedSections(0);
+        } else if (state.selectedSectionIndexes.start !== state.selectedSectionIndexes.end) {
+          updateSelectedSections(state.selectedSectionIndexes.end);
+        } else if (state.selectedSectionIndexes.start < state.sections.length - 1) {
+          updateSelectedSections(state.selectedSectionIndexes.start + 1);
+        }
```

Comment:
Switching these cases solves the [focus management issue](https://github.com/mui/mui-x/pull/5504#issuecomment-1221503822).
```suggestion
        } else if (state.selectedSectionIndexes.start < state.sections.length - 1) {
          updateSelectedSections(state.selectedSectionIndexes.start + 1);
        } else if (state.selectedSectionIndexes.start !== state.selectedSectionIndexes.end) {
          updateSelectedSections(state.selectedSectionIndexes.end);
        } 
```